omero/developers/testing.txt

```diff
@@ -491,3 +374,196 @@ add paths to :envvar:`PYTHONPATH` specifically for `pytest`.
+
+Writing tests
+-------------
+
+Writing Java tests
+^^^^^^^^^^^^^^^^^^
+
+For more information on writing tests in general see `<http://testng.org>`_.
+For a test to be an "integration" test, place it in the "integration"
+TestNG group. If a test is temporarily broken, add it to the "broken" group:
+
+::
+
+    @Test(groups = {"integration", "broken"}
+    public void testMyStuff() {
+
+    }
+
+Tests should be of the **Acceptance Test** form. The ticket number
+for which a test is being written should be added in the TestNG annotation:
+
+::
+
+    @Test(groups = "ticket:60")
+
+This works at either the method level (see :source:`SetsAndLinksTest.java
+<components/model/test/ome/model/utests/SetsAndLinksTest.java>`)
+or the class level (see :source:`UniqueResultTest.java
+<components/server/test/ome/server/itests/query/UniqueResultTest.java>`).
+
+The tests under :sourcedir:`components/tools/OmeroJava/test` will be the
+starting point for most Java-client developers coming to OMERO. An example
+skeleton for an integration test looks similar to
+
+::
+
+    @Test(groups = "integration")
+    public class MyTest {
+
+      omero.client client;
+
+      @BeforeClass
+      protected void setup() throws Exception {
+        client = new omero.client();
+        client.createSession();
+      }
+
+      @AfterClass
+      protected void tearDown() throws Exception {
+        client.closeSession();
+      }
+
+      @Test
+      public void testSimple() throws Exception {
+        client.getSession().getAdminService().getEventContext();
+      }
+
+    }
+
+
+.. _writing-python-tests:
+
+Writing Python tests
+^^^^^^^^^^^^^^^^^^^^
+
+To write and run Python tests you first need to install `pytest`:
+
+::
+
+    pip install pytest
+
+For more information on writing tests in general see `<http://pytest.org>`_.
+
+Similar to the OmeroJava tests, the tests under
+:sourcedir:`components/tools/OmeroPy/test`,
+:sourcedir:`components/tools/OmeroFS/test` and
+:sourcedir:`components/tools/OmeroWeb/test` will be the starting point
+for most Python-client developers coming to OMERO. Integration tests should
+be placed under the :file:`integration` subfolders. The file names must begin
+with `test_` for the tests to be found by `pytest`.
+
+::
+
+    import omero
+
+    class TestExample(object)
+
+      def setup_method(self, method):
+        client = new omero.client()
+        client.createSession()
+
+      def teardown_method(self, method):
+        client.closeSession()
+
+      def testSimple():
+        ec = client.getSession().getAdminService().getEventContext()
+        assert ec, "No EventContext!"
+
+
+Using the Python test library
+"""""""""""""""""""""""""""""
+
+The :source:`OMERO Python test library <components/tests/python/library/__init__.py>`
+defines an abstract ``ITest`` class that implements the connection set up as
+well as many functionalities shared amongst all Python integration tests.
```

Comment:
"many functionalities" -> "much functionality"  - I don't think functionality is comfortable as a countable noun.